src/driver-sinowealth.c

```diff
@@ -0,0 +1,475 @@
+/*
+ * Copyright © 2020 Marian Beermann
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "libratbag-private.h"
+#include "libratbag-hidraw.h"
+
+#define SINOWEALTH_REPORT_ID_CONFIG 0x4
+#define SINOWEALTH_REPORT_ID_CMD 0x5
+#define SINOWEALTH_CMD_FIRMWARE_VERSION 0x1
+#define SINOWEALTH_CMD_GET_CONFIG 0x11
+#define SINOWEALTH_CONFIG_SIZE 520
+
+#define SINOWEALTH_XY_INDEPENDENT 0x80
+
+/* The PC software only goes down to 400, but the PMW3360 doesn't care */
+#define SINOWEALTH_DPI_MIN 100
+#define SINOWEALTH_DPI_MAX 12000
+#define SINOWEALTH_DPI_STEP 100
+
+/* other models might have up to eight */
+#define SINOWEALTH_NUM_DPIS 6
+
+#define SINOWEALTH_RGB_BRIGHTNESS_BITS 0xF0
+#define SINOWEALTH_RGB_SPEED_BITS 0x0F
+
+struct RGB8 {
+	uint8_t r, g, b;
+} __attribute__((packed));
+
+_Static_assert(sizeof(struct RGB8) == 3, "Invalid size");
+
+struct RBG8 {
+	uint8_t r, b, g;
+} __attribute__((packed));
+
+_Static_assert(sizeof(struct RBG8) == 3, "Invalid size");
+
+enum rgb_effect {
+	RGB_OFF = 0,
+	RGB_GLORIOUS = 0x1,   /* unicorn mode */
```

Comment:
Maybe a better description here. If find neither `RGB_GLORIOUS` nor `unicorn mode` very descriptive.

Reply:
All these different mice have different color modes, I don't see the necessity to describe them all, especially since we don't support them.

Reply:
"Glorious mode" is what their manual/software calls it. It looks like a rainbow cycling around the mouse.